exercises/react/src/example.c

```diff
@@ -0,0 +1,210 @@
+{
+   struct child *child = r->first;
+   while (child) {
+      struct cb *cb = child->cell->cb;
+      while (cb) {
+         struct cb *next_cb = cb->next;
+         free(cb);
+         cb = next_cb;
+      }
+      free(child->cell);
+
+      struct child *next = child->next;
+      free(child);
+      child = next;
+   }
+   free(r);
+}
+
+#define SUCCESS 1
+#define FAIL 0
+
+static int add_child(struct reactor *r, struct cell *cell)
+{
+   struct child *child = calloc(1, sizeof(struct child));
+   if (!child) {
+      return FAIL;
+   }
+   child->cell = cell;
+   if (!r->first) {
+      r->first = child;
+   } else {
+      r->last->next = child;
+   }
+   r->last = child;
+   return SUCCESS;
+}
+
+struct cell *create_input_cell(struct reactor *r, int initial_value)
+{
+   struct cell *c = calloc(1, sizeof(struct cell));
+   if (!c) {
+      return NULL;
+   }
+   if (add_child(r, c) != SUCCESS) {
+      free(c);
+      return NULL;
+   }
+   c->reactor = r;
+   c->kind = kind_input;
+   c->value = initial_value;
+   return c;
+}
+
+struct cell *create_compute1_cell(struct reactor *r, struct cell *input,
+                                  compute1 compute)
+{
+   struct cell *c = calloc(1, sizeof(struct cell));
+   if (!c) {
+      return NULL;
+   }
+   if (add_child(r, c) != SUCCESS) {
+      free(c);
+      return NULL;
+   }
+   c->reactor = r;
+   c->kind = kind_compute1;
+   c->input1 = input;
+   c->compute1 = compute;
+   c->value = compute(cell_value(input));
+   return c;
+}
+
+struct cell *create_compute2_cell(struct reactor *r, struct cell *input1,
+                                  struct cell *input2, compute2 compute)
+{
+   struct cell *c = calloc(1, sizeof(struct cell));
+   if (!c) {
+      return NULL;
+   }
+   if (add_child(r, c) != SUCCESS) {
+      free(c);
+      return NULL;
+   }
+   c->reactor = r;
+   c->kind = kind_compute2;
+   c->input1 = input1;
+   c->input2 = input2;
+   c->compute2 = compute;
+   c->value = compute(cell_value(input1), cell_value(input2));
+   return c;
+}
+
+int cell_value(struct cell *c)
+{
+   return c->value;
+}
+
+static void propagate(struct cell *c)
+{
+   int new_value;
+   switch (c->kind) {
+   case kind_compute1:
+      new_value = c->compute1(cell_value(c->input1));
+      break;
+   case kind_compute2:
+      new_value = c->compute2(cell_value(c->input1), cell_value(c->input2));
+      break;
+   default:
+      return;
+   }
+
+   if (new_value != c->value) {
+      c->value = new_value;
+      for (struct cb * cb = c->cb; cb; cb = cb->next) {
+         cb->f(cb->obj, new_value);
+      }
+   }
+
+}
+
+void cell_value_set(struct cell *c, int new_value)
+{
+   c->value = new_value;
+   struct reactor *r = c->reactor;
+
+   // We take the very naive route of updating all cells.
+   // Traversing the tree and only updating the cells dependent on the just-changed cell is possible,
+   // but requires much more memory management to make each cell aware of its dependents.
```

Comment:
I think this is really against the spirit of reactive programming, but it is a pretty elegant (if non-performant) solution. If we had more of the basic building blocks available to us (like STL) then I'd say we should definitely target a more robust solution, but I feel like this would quickly spin out of control in no-dependencies C.

Thoughts?